rfcs/20220309-feature_flags.md

```diff
@@ -0,0 +1,303 @@
+        "release_date": "2023-01-01"
+      }
+    },
+    "feature_name_2": {
+      "status": "switchable",
+      "default": "on", // whether the feature is opt-in or opt-out
+      "description": "Feature 2 does ...",
+    }
+  }
+}
+```
+
+Conf files being JSON can be read in any part of the codebase, without any additional libraries. Note that conf files only define the status of the flag and not the final state of the feature.
+
+#### Cookies
+
+For flags that are set to switchable in the conf files, we use the state in the cookie to determine whether the flag is set to on or off. This cookie has the following schema.
+
+```json
+{
+  "features": {
+    "feature_name_1": "on",
+    "feature_name_2": "off"
+  }
+}
+```
+
+The cookie is only referred to for flags that are set to switchable, any other feature definitions in the cookie should be dropped or ignored. If the cookie does not mention the flag, the `default` value from the config is read.
+
+Regardless of the storage mechanism we will need a few other key things.
+
+### Store
+
+To prevent feature flags from being all over the place, we can keep them store in a Pinia store and just use the store wherever there is a check. This allows us to toggle features from a centralised place.
+
+Since keys will be added and removed periodically, the store must not expose feature flags directly but rather expose a getter that can handle fallback-to-default.
+
+The flags can be populated in the store from whatever source we choose.
+
+### Utilities
+
+A composable can be provided that interacts with the Pinia store to provide `isFeatureEnabled` functionality. The composable can directly be pulled into a component and used as a boolean `ref`.
+
+### Directive
+
+To prevent a lot of `v-if="isFeatureEnabled(feature)"` statements in the templates, we can create a new directive.
+
+```
+v-flag:{feature_name: str}="{flag_state: str}"
+```
+
+This directive takes the `feature_name: str` argument and the `flag_state: boolean` value. The component is only rendered when the state of the feature flag matches the given value.
+
+The component is shown if the flag is
+
+- enabled
+- switchable (with the user having chosen to enable it)
+
+It internally intefaces with the utility to evaluate whether the component is rendered or not. Here is [some code](https://github.com/mblarsen/vue-browser-acl/blob/4e3bb90d2ba4fcc3edd30b27737f4531dc464329/index.js#L136-L163) that can be used to avoid rendering.
+
+### Drawbacks
+
+This implementation has the main drawback of leaving `v-flag` directives around the codebase that are equivalent `v-if="true"` when the feature has been permanently enabled. Similarly a script utilitiy `isFeatureEnabled` that always evaluates to `true` is another a no-op.
+
+We can peridically clean them up or create tickets to remove them after a feature is completely and irreversibly live. This is important to prevent cruft from accumulating in in the codebase.
```

Comment:
```suggestion
We can periodically clean them up or create tickets to remove them after a feature is completely and irreversibly live. This is important to prevent cruft from accumulating in in the codebase.
```